File: src/services/pitch/PitchDetectionService.ts
/**
 * Pitch Detection Service
 * --------------------------------------------------------
 * Central real-time DSP engine that performs:
 *   - YIN fundamental frequency estimation
 *   - Zero-crossing fallback (helps on low strings)
 *   - Confidence gating
 *   - Frequency range validation
 *   - Frequency → musical note conversion
 *
 * IMPORTANT:
 * This service contains NO audio I/O and NO onset / silence detection.
 * It assumes:
 *   - Input buffers are already gated and aligned by the native layer
 *   - Buffers are fixed-size NOTE windows (e.g. 2048 samples)
 */

import { YinPitchDetector } from '@/utils/audio/yin';
import { FFT, applyHannWindow } from '@/utils/audio/fft';
import { frequencyToNote, GUITAR_FREQUENCY_RANGE } from '@/utils/music';
import { AudioDetection } from '@/types';

export interface PitchDetectionConfig {
  sampleRate: number;
  bufferSize: number;      // NOTE window size (native guarantees this)
  yinThreshold: number;
  minConfidence: number;
  minFrequency: number;
  maxFrequency: number;
}

/**
 * Default configuration for guitar-range pitch detection
 */
export const DEFAULT_PITCH_CONFIG: PitchDetectionConfig = {
  sampleRate: 44100,
  bufferSize: 2048,                 // NOTE window (native side)
  yinThreshold: 0.15,
  minConfidence: 0.6,
  minFrequency: GUITAR_FREQUENCY_RANGE.MIN - 10,  // margin below E2
  maxFrequency: GUITAR_FREQUENCY_RANGE.MAX + 100, // margin above high frets
};

export type PitchDetectionCallback = (d: AudioDetection | null) => void;

// Debug flag
const DEBUG_PITCH = false;

function debugLog(...args: any[]) {
  if (DEBUG_PITCH) {
    console.log(...args);
  }
}

export class PitchDetectionService {
  private config: PitchDetectionConfig;

  // Core DSP tools
  private yinDetector: YinPitchDetector;
  private fft: FFT; // reserved for future chord detection

  // Runtime state
  private detectionCallback: PitchDetectionCallback | null = null;
  private isActive = false;

  constructor(config: PitchDetectionConfig = DEFAULT_PITCH_CONFIG) {
    this.config = config;

    this.yinDetector = new YinPitchDetector(
      config.sampleRate,
      config.yinThreshold,
      config.bufferSize
    );

    this.fft = new FFT(config.bufferSize);
  }

  /**
   * Activates pitch detection.
   * The service expects already-gated NOTE windows from the pipeline.
   */
  start(callback: PitchDetectionCallback): void {
    this.detectionCallback = callback;
    this.isActive = true;
    console.log('[PitchDetection] started');
  }

  /**
   * Deactivates detection.
   */
  stop(): void {
    this.isActive = false;
    this.detectionCallback = null;
    console.log('[PitchDetection] stopped');
  }

  /**
   * Main processing entry point.
   * Consumes a NOTE-sized PCM frame and attempts to detect pitch.
   */
  processSamples(samples: Float32Array, timestampMs: number): AudioDetection | null {
    if (!this.isActive) return null;

    // Native guarantees correct NOTE window size
    const buffer = samples;

    /**
     * STEP 1 — (Optional) Windowing
     * Hann window is applied for spectral consistency and future FFT usage.
     */
    const windowed = applyHannWindow(buffer);

    /**
     * STEP 2 — Primary algorithm: YIN
     */
    let yinResult = this.yinDetector.detect(buffer);

    if (yinResult) {
      debugLog(
        '[YIN]',
        'freq:', yinResult.frequency.toFixed(2),
        'conf:', yinResult.confidence.toFixed(3),
        'per:', yinResult.periodicity?.toFixed(3)
      );
    } else {
      debugLog('[YIN] null');
    }

    /**
     * STEP 3 — Fallback for low strings (E2, A2)
     * Zero-crossing gives a rough but usable frequency estimate.
     */
    if (!yinResult) {
      const fallback = this.estimateFrequencyZeroCrossing(
        windowed,
        this.config.sampleRate
      );

      if (!fallback) {
        this.detectionCallback?.(null);
        return null;
      }

      yinResult = {
        frequency: fallback,
        confidence: 0.85,
        periodicity: 0.5,
      };
    }

    /**
     * STEP 4 — Confidence gating
     * Low notes are allowed slightly weaker confidence.
     */
    const isLow = yinResult.frequency < 110; // E2–D3 region
    const minConf = isLow ? 0.5 : this.config.minConfidence;

    if (yinResult.confidence < minConf) {
      debugLog('[CONF] gated', yinResult.confidence.toFixed(3));
      this.detectionCallback?.(null);
      return null;
    }

    /**
     * STEP 5 — Expected frequency range check
     */
    if (
      yinResult.frequency < this.config.minFrequency ||
      yinResult.frequency > this.config.maxFrequency
    ) {
      debugLog('[RANGE] gated', yinResult.frequency.toFixed(2));
      this.detectionCallback?.(null);
      return null;
    }

    /**
     * STEP 6 — Convert frequency → musical note
     */
    const note = frequencyToNote(
      yinResult.frequency,
      yinResult.confidence
    );

    /**
     * FINAL — Build detection result
     */
    const detection: AudioDetection = {
      timestamp: timestampMs,
      frequency: yinResult.frequency,
      amplitude: 0, // amplitude is now owned by native layer
      confidence: yinResult.confidence,
      note,
    };

    this.detectionCallback?.(detection);
    return detection;
  }

  /**
   * Zero-Crossing estimation
   * Lightweight and effective for fundamental-heavy low notes.
   */
  private estimateFrequencyZeroCrossing(
    samples: Float32Array,
    sampleRate: number
  ): number | null {
    let crossings = 0;
    let prev = samples[0];

    for (let i = 1; i < samples.length; i++) {
      const curr = samples[i];
      if ((prev <= 0 && curr > 0) || (prev >= 0 && curr < 0)) {
        crossings++;
      }
      prev = curr;
    }

    const duration = samples.length / sampleRate;
    if (duration <= 0 || crossings < 2) return null;

    return crossings / (2 * duration);
  }

  /**
   * Update sample rate if native engine changes it.
   */
  setSampleRate(sampleRate: number): void {
    if (!sampleRate || sampleRate <= 0) return;
    if (sampleRate === this.config.sampleRate) return;

    this.config.sampleRate = sampleRate;
    this.yinDetector.setSampleRate(sampleRate);

    console.log('[PitchDetection] sample rate updated →', sampleRate);
  }

  isRunning(): boolean {
    return this.isActive;
  }
}

/**
 * Global singleton instance
 */
let pitchDetectionServiceInstance: PitchDetectionService | null = null;

export function getPitchDetectionService(config?: PitchDetectionConfig) {
  if (!pitchDetectionServiceInstance) {
    pitchDetectionServiceInstance = new PitchDetectionService(config);
  }
  return pitchDetectionServiceInstance;
}

export function resetPitchDetectionService() {
  if (pitchDetectionServiceInstance) {
    pitchDetectionServiceInstance.stop();
  }
  pitchDetectionServiceInstance = null;
}

File: ios/PitchDetection/PitchDetectionService.swift
import Foundation

// MARK: - PitchDetectionService
// --------------------------------------------------
// JS parity implementation.
// Mirrors services/audio/PitchDetectionService.ts
//
// Responsibilities:
// - YIN pitch detection (RAW buffer)
// - Zero-crossing fallback (HANN windowed)
// - Confidence gating
// - Frequency range gating
// - Frequency → Note conversion
//
// IMPORTANT:
// - No RMS / amplitude gate here
// - No onset / silence logic here
// - No smoothing filter here
// --------------------------------------------------

final class PitchDetectionService {

    // MARK: - Config & State

    private var config: PitchDetectionConfig
    private var isActive: Bool = false

    // Core DSP
    private let yin: YinPitchDetector

    // MARK: - Init

    init(config: PitchDetectionConfig = PitchDefaults.DEFAULT_CONFIG) {
        self.config = config
        self.yin = YinPitchDetector(
            sampleRate: config.sampleRate,
            threshold: config.yinThreshold,
            bufferSize: config.bufferSize
        )
    }

    // MARK: - Lifecycle

    func start() {
        isActive = true
        print("[PitchDetection] started")
    }

    func stop() {
        isActive = false
        print("[PitchDetection] stopped")
    }

    func isRunning() -> Bool {
        return isActive
    }

    func setSampleRate(_ sampleRate: Double) {
        guard sampleRate > 0 else { return }
        guard sampleRate != config.sampleRate else { return }

        config.sampleRate = sampleRate
        yin.setSampleRate(sampleRate)

        print("[PitchDetection] sample rate updated → \(sampleRate)")
    }

    // MARK: - Main Processing (JS parity)

    /// NOTE-sized buffer expected (native guarantees size)
    func processSamples(
        samples: [Float],
        timestampMs: Int,
        amplitude: Double
    ) -> [String: Any]? {

        guard isActive else { return nil }
        guard samples.count == config.bufferSize else { return nil }

        // --------------------------------------------------
        // STEP 1 — Primary: YIN (RAW buffer, no window)
        // --------------------------------------------------
        var yinResult = yin.detect(samples: samples)

        // --------------------------------------------------
        // STEP 2 — Fallback: Zero-crossing
        // JS parity: uses HANN-windowed signal
        // --------------------------------------------------
        if yinResult == nil {
            let windowed = applyHannWindow(samples)

            guard let fallbackFreq = estimateFrequencyZeroCrossing(
                samples: windowed,
                sampleRate: config.sampleRate
            ) else {
                return nil
            }

            yinResult = YinResult(
                frequency: fallbackFreq,
                confidence: 0.85,
                periodicity: 0.5
            )
        }

        guard let result = yinResult else { return nil }

        // --------------------------------------------------
        // STEP 3 — Confidence gating
        // Low notes (E2–G3, below 200Hz) allowed weaker confidence
        // --------------------------------------------------
        let isLowFreq = result.frequency < 200.0  // Covers E2, A2, D3, G3
        let minConf = isLowFreq ? 0.40 : config.minConfidence

        guard result.confidence >= minConf else {
            return nil
        }

        // --------------------------------------------------
        // STEP 4 — Frequency range gating
        // --------------------------------------------------
        guard
            result.frequency >= config.minFrequency,
            result.frequency <= config.maxFrequency
        else {
            return nil
        }

        // --------------------------------------------------
        // STEP 5 — Frequency → Note
        // --------------------------------------------------
        let note = MusicTheory.frequencyToNote(
            result.frequency,
            confidence: result.confidence
        )

        // --------------------------------------------------
        // FINAL — Payload
        // --------------------------------------------------
        return AudioDetectionPayload.note(
            timestampMs: timestampMs,
            frequency: result.frequency,
            amplitude: amplitude,
            confidence: result.confidence,
            note: note
        )
    }

    // MARK: - Zero Crossing (JS parity)

    private func estimateFrequencyZeroCrossing(
        samples: [Float],
        sampleRate: Double
    ) -> Double? {

        guard samples.count > 1 else { return nil }
        guard sampleRate > 0 else { return nil }

        var crossings = 0
        var prev = samples[0]

        for i in 1..<samples.count {
            let curr = samples[i]
            if (prev <= 0 && curr > 0) || (prev >= 0 && curr < 0) {
                crossings += 1
            }
            prev = curr
        }

        let duration = Double(samples.count) / sampleRate
        guard duration > 0 else { return nil }
        guard crossings >= 2 else { return nil }

        return Double(crossings) / (2.0 * duration)
    }
}
